File: diabetes.py
import pandas as pd
import numpy as np


#loading the dataset
df = pd.read_csv("D:/BLR10AM/Assi/16.Decision tree/Datasets_DTRF/Diabetes.csv")


#2.	Work on each feature of the dataset to create a data dictionary as displayed in the below image
#######feature of the dataset to create a data dictionary

#######feature of the dataset to create a data dictionary


d_types =["count","ratio","ratio","ratio","ratio","ratio","ratio","ratio","binary"]

data_details =pd.DataFrame({"column name":df.columns,
                            "data types ":d_types,
                            "data types-p":df.dtypes})


            #3.	Data Pre-processing
          #3.1 Data Cleaning, Feature Engineering, etc
          
          
#details of df 
df.info()
df.describe()          


#data types        
df.dtypes


#checking for na value
df.isna().sum()
df.isnull().sum()
df.dropna()

#checking unique value for each columns
df.nunique()

#variance of df
df.var()



"""4.	Exploratory Data Analysis (EDA):
4.1.	Summary
4.2.	Univariate analysis
4.3.	Bivariate analysis
	 """
    


EDA ={"column ": df.columns,
      "mean": df.mean(),
      "median":df.median(),
      "mode":df.mode(),
      "standard deviation": df.std(),
      "variance":df.var(),
      "skewness":df.skew(),
      "kurtosis":df.kurt()}

EDA

# covariance for data set 
covariance = df.cov()
covariance


####### graphical repersentation 

##historgam and scatter plot
import seaborn as sns
sns.pairplot(df.iloc[:, :])


df.nunique()



colnames = list(df.columns)

predictors = colnames[0:8]
target = colnames[8]

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size = 0.2,random_state=7)

"""
5.	Model Building
5.1	Build the model on the scaled data (try multiple options)
5.2	Perform Decision Tree and Random Forest on the given datasets.
5.3	Train and Test the data and perform cross validation techniques, compare accuracies, precision and recall and explain about them.
5.4	Briefly explain the model output in the documentation. """


from sklearn.tree import DecisionTreeClassifier as DT


model = DT(criterion = 'entropy')
model.fit(train[predictors], train[target])


# Prediction on Test Data
preds = model.predict(test[predictors])
pd.crosstab(test[target], preds, rownames=['Actual'], colnames=['Predictions'])

np.mean(preds == test[target]) # Test Data Accuracy 

# Prediction on Train Data
preds = model.predict(train[predictors])
pd.crosstab(train[target], preds, rownames = ['Actual'], colnames = ['Predictions'])

np.mean(preds == train[target]) # Train Data Accuracy

#model is over fitting so we are building random forest

###########  Random forest 



from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

rf_clf.fit(train[predictors], train[target])

from sklearn.metrics import accuracy_score, confusion_matrix

confusion_matrix(test[target], rf_clf.predict(test[predictors]))
accuracy_score(test[target], rf_clf.predict(test[predictors]))


######
# GridSearchCV

from sklearn.model_selection import GridSearchCV

rf_clf_grid = RandomForestClassifier(n_estimators=500, n_jobs=1, random_state=42)

param_grid = {"max_features": [4, 5, 6, 7, 8, 9, 10], "min_samples_split": [2, 3, 10]}

grid_search = GridSearchCV(rf_clf_grid, param_grid, n_jobs = -1, cv = 5, scoring = 'accuracy')

grid_search.fit(train[predictors], train[target])

grid_search.best_params_

cv_rf_clf_grid = grid_search.best_estimator_

from sklearn.metrics import accuracy_score, confusion_matrix

confusion_matrix(test[target], cv_rf_clf_grid.predict(test[predictors]))
accuracy_score(test[target], cv_rf_clf_grid.predict(test[predictors]))

File: diabetes.R

library(party)

library(caret)

library(C50)
library(tree)
library(gmodels)

df <- read.csv(file.choose())

##Exploring and preparing the data ----
str(df)

names(df)[9] <- paste("Class_variable")
# Splitting data into training and testing.
# splitting the data based on Sales
df$Class_variable <- as.factor(df$Class_variable)


# Shuffle the data
CD_ <- df[order(runif(768)), ]
str(CD_)

CD_train <- CD_[1:550,]

# View(CD_train)
CD_test <- CD_[551:768,]

# View(CD_test)


# Step 3: Training a model on the data

library(C50)

credit_model <- C5.0(CD_train[, -9], CD_train$Class_variable)

windows()
plot(credit_model) 

# Display detailed information about the tree
summary(credit_model)

# Step 4: Evaluating model performance
# Test data accuracy
test_res <- predict(credit_model, CD_test[, -9])
test_acc <- mean(CD_test$Class_variable == test_res)
test_acc

# cross tabulation of predicted versus actual classes
library(gmodels)
CrossTable(CD_test$Class_variable, test_res, dnn = c('actual default', 'predicted default'))

# On Training Dataset
train_res <- predict(credit_model, CD_train[, -9])
train_acc <- mean(CD_train$Class_variable == train_res)
train_acc

table(CD_train$Class_variable, train_res)
